File: app/src/main/java/com/example/app/jasper/routinedeveloper_v2/repository/TodoListRepository.java
package com.example.app.jasper.routinedeveloper_v2.repository;

import android.content.Context;
import android.util.Log;

import com.example.app.jasper.routinedeveloper_v2.model.SQLDatabaseHelper;
import com.example.app.jasper.routinedeveloper_v2.model.Todo;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.List;

public class TodoListRepository {

    private static TodoListRepository instance;
    private SQLDatabaseHelper dataBase;
    private SharedPreferenceHelper prefs;

    public static TodoListRepository getInstance(Context context) {
        if (instance == null) {
            instance = new TodoListRepository(context);
        }
        return instance;
    }

    public TodoListRepository(Context context) {
        dataBase = SQLDatabaseHelper.getInstance(context);
        prefs = SharedPreferenceHelper.INSTANCE;
    }

    public List<Todo> getAllItems() {
        return dataBase.readAllItems();
    }

    public long createItem(Todo item) {
        return dataBase.createItem(item);
    }

    public void updateItem(long id, Todo item) {
        dataBase.updateItem(id, item);
    }

    public void deleteItem(long id) {
        dataBase.deleteItem(id);
    }

    public List<Todo> readAllItems() {
        return dataBase.readAllItems();
    }

    public Todo readItem(long id) {
        return dataBase.readItem(id);
    }



    public void clearTargetDate() {
        prefs.setChallengeEndingDate("");
    }

    public void clearScore() {
        prefs.setUndoneCount(0);
        prefs.setDoneCount(0);
    }

    public int getStoredDay() {
        return prefs.getStoredDay();
    }

    public void setCurrentDay(int currentDay) {
        prefs.updateDate(currentDay);
    }

    public void incrementDoneCounter() {
        int doneCount = prefs.getDoneCount();
        doneCount += 1;
        prefs.setDoneCount(doneCount);
        Log.d("COUNTER", "doneCount incremented");


        int done = prefs.getDoneCount();
        Log.d("COUNTER", "new done value: " + String.valueOf(done));

    }

    public void incrementUndoneCounter() {
        int undoneCount = prefs.getUndoneCount();
        undoneCount += 1;
        prefs.setUndoneCount(undoneCount);
        Log.d("COUNTER", "undoneCount incremented");


        int undone = prefs.getUndoneCount();
        Log.d("COUNTER", "new undone value: " + undone);
    }

    public void setChallengeEndingDate(String endingDate) {
        prefs.setChallengeEndingDate(endingDate);
    }

    public void setFirstStart(boolean b) {
        prefs.setFirstStart(b);
    }

    public void saveList(List<Todo> todos) {
        for (Todo item : todos)
            dataBase.updateItem(item.getId(), item);
    }

    public void addItem(Todo item) {
        dataBase.updateItem(item.getId(), item);
    }

    @Nullable
    public Integer getUndoneCount() {
        return prefs.getUndoneCount();
    }

    @Nullable
    public Integer getDoneCount() {
        return prefs.getDoneCount();
    }

    @Nullable
    public String getChallengeEndingDate() {
        return prefs.getChallengeEndingDate();
    }
}

File: app/src/main/java/com/example/app/jasper/routinedeveloper_v2/repository/TodoListRepository.kt
package com.example.app.jasper.routinedeveloper_v2.repository

import android.content.Context
import android.util.Log
import com.example.app.jasper.routinedeveloper_v2.model.SQLDatabaseHelper
import com.example.app.jasper.routinedeveloper_v2.model.Todo

class TodoListRepository(context: Context) {
    private val dataBase: SQLDatabaseHelper = SQLDatabaseHelper.getInstance(context)
    private val prefs: SharedPreferenceHelper = SharedPreferenceHelper
    val allItems: List<Todo>
        get() = dataBase.readAllItems()

    fun createItem(item: Todo?): Long {
        return dataBase.createItem(item)
    }

    suspend fun updateItem(id: Long, item: Todo?) {
        dataBase.updateItem(id, item)
    }

    suspend fun deleteItem(id: Long) {
        dataBase.deleteItem(id)
    }

    suspend fun readAllItems(): List<Todo> {
        return dataBase.readAllItems()
    }

    suspend fun readItem(id: Long): Todo {
        return dataBase.readItem(id)
    }

    fun clearTargetDate() {
        prefs.challengeEndingDate = ""
    }

    fun clearScore() {
        prefs.overallUndoneCount = 0
        prefs.overallDoneCount = 0
    }

    val storedDay: Int
        get() = prefs.storedDay

    fun setCurrentDay(currentDay: Int) {
        prefs.updateDate(currentDay)
    }

    fun incrementOverallDoneCounter() {
        var doneCount = prefs.overallDoneCount
        doneCount += 1
        prefs.overallDoneCount = doneCount
        Log.d("COUNTER", "doneCount incremented")
        val done = prefs.overallDoneCount
        Log.d("COUNTER", "new done value: $done")
    }

    fun incrementOverallUndoneCounter() {
        var undoneCount = prefs.overallUndoneCount
        undoneCount += 1
        prefs.overallUndoneCount = undoneCount
        Log.d("COUNTER", "undoneCount incremented")
        val undone = prefs.overallUndoneCount
        Log.d("COUNTER", "new undone value: $undone")
    }

    fun setFirstStart(b: Boolean) {
        prefs.firstStart = b
    }

    suspend fun saveList(todos: List<Todo>) {
        for (item in todos) dataBase.updateItem(item.id, item)
    }

    suspend fun addItem(item: Todo) {
        dataBase.updateItem(item.id, item)
    }

    val undoneCount: Int?
        get() = prefs.overallUndoneCount

    val doneCount: Int?
        get() = prefs.overallDoneCount

    var challengeEndingDate: String?
        get() = prefs.challengeEndingDate
        set(endingDate) {
            prefs.challengeEndingDate = endingDate!!
        }

    companion object {
        private var instance: TodoListRepository? = null
        fun getInstance(context: Context): TodoListRepository {
            if (instance == null) {
                instance = TodoListRepository(context)
            }
            return instance as TodoListRepository
        }
    }

}